### PR TITLE
Polynomial cleanup: mem allocation and raw pointers

### PR DESCRIPTION
**Pull request by ledwards2225**
# Description

- Create single function for memory allocation in Polynomial
- Replace raw pointers 
- Remove ASSERT from operator[]
- TBD

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

File: cpp/src/aztec/plonk/proof_system/prover/prover.cpp [hunk 1]
```diff
@@ -665,15 +665,10 @@ template <typename settings> void ProverBase<settings>::add_plookup_memory_recor
...
-    // fr* w_2 = key->polynomial_cache.get("w_2_fft").get_coefficients();
-    // fr* w_3 = key->polynomial_cache.get("w_3_fft").get_coefficients();
-    fr* w_1 = key->polynomial_cache.get("w_1_lagrange").get_coefficients();
-    fr* w_2 = key->polynomial_cache.get("w_2_lagrange").get_coefficients();
-    fr* w_3 = key->polynomial_cache.get("w_3_lagrange").get_coefficients();
+
+    std::span<fr> w_1 = key->polynomial_cache.get("w_1_lagrange");
+    std::span<fr> w_2 = key->polynomial_cache.get("w_2_lagrange");
+    std::span<fr> w_3 = key->polynomial_cache.get("w_3_lagrange");
     fr* w_4 = key->polynomial_cache.get("w_4_lagrange").get_coefficients();
```
Comment (adr1anh) [inline cpp/src/aztec/plonk/proof_system/prover/prover.cpp]: Why is this not a span

File: cpp/src/aztec/honk/composer/composer_helper/composer_helper.cpp [hunk 2]
```diff
@@ -97,12 +97,12 @@ std::shared_ptr<waffle::verification_key> ComposerHelper<CircuitConstructor>::co
             poly_info.source == waffle::PolynomialSource::OTHER) {
             // Fetch the polynomial in its vector form.
 
-            fr* poly_coefficients;
-            poly_coefficients = proving_key->polynomial_cache.get(poly_label).get_coefficients();
+            std::span<fr> poly_coefficients;
+            poly_coefficients = proving_key->polynomial_cache.get(poly_label);
```
Comment (adr1anh) [inline cpp/src/aztec/honk/composer/composer_helper/composer_helper.cpp]: combine

File: cpp/src/aztec/plonk/proof_system/widgets/random_widgets/permutation_widget_impl.hpp [hunk 3]
```diff
@@ -380,22 +377,22 @@ barretenberg::fr ProverPermutationWidget<program_width, idpolys, num_roots_cut_o
     // (w_l(X) + β.σ_1(X) + γ).(w_r(X) + β.σ_2(X) + γ).(w_o(X) + β.σ_3(X) + γ).z(X).α
     // Once we divide by the vanishing polynomial, this will be a degree 3n polynomial. (4 * (n-1) - (n-4)).
 
-    std::array<const fr*, program_width> wire_ffts;
-    std::array<const fr*, program_width> sigma_ffts;
-    [[maybe_unused]] std::array<const fr*, program_width> id_ffts;
+    std::array<std::span<fr>, program_width> wire_ffts;
```
Comment (adr1anh) [inline cpp/src/aztec/plonk/proof_system/widgets/random_widgets/permutation_widget_impl.hpp]: can do `std::span<const fr>`

File: cpp/src/aztec/plonk/proof_system/prover/prover.cpp (hunk 1 above)
Comment (ledwards2225) [reply cpp/src/aztec/plonk/proof_system/prover/prover.cpp]: The wire `w_4` is modified just below this and I was following a convention that spans are 'read only'. I added const to the other three and changed this to a span.

File: cpp/src/aztec/honk/composer/composer_helper/composer_helper.cpp (hunk 2 above)
Comment (ledwards2225) [reply cpp/src/aztec/honk/composer/composer_helper/composer_helper.cpp]: Removed these lines altogether

File: cpp/src/aztec/plonk/proof_system/widgets/random_widgets/permutation_widget_impl.hpp (hunk 3 above)
Comment (ledwards2225) [reply cpp/src/aztec/plonk/proof_system/widgets/random_widgets/permutation_widget_impl.hpp]: Added `const` here and in many other locations

**Review by adr1anh — APPROVED** [review]
LGTM

**Pull request merged**